### ISSUE TITLE
GenerateDungeonMazeMap map generate disconnected parts of the maze

### ISSUE THREAD

**Issue by Chris3606** [Chris3606/GoRogue#85]
Currently, tunnels of the maze may be generated that are completely separated from the rest of the map by walls.

Currently,  this is because the connection algorithm only iterates over the sides of _rooms_ looking for connection locations.  Because there can be multiple disconnected sections of mazes generated by `MazeGenerator` (because the algorithm strives to fill the entire free space with mazes), the original article upon which the algorithm is based treated those mazes as "regions", just like the rooms.  This is one possibility for a fix -- the connection algorithm would effectively run the `MapAreaFinder`, then, for each area (rectangular and non-rectangular), find connection points and connect them together.

In context of tunnels, this shouldn't be too difficult.  Another option, however, that I have not tested but that in theory would prove more versatile in the case where the areas are more arbitrarily-shaped:

This could be accomplished by an algorithm:
- Take map with mazes and rooms.  Run `MapAreaFinder` to obtain `areas`.
- Create `DisjointSet` for each area.
- for each `area` in `areas`
    - for each `point` in area locations `//This is way overkill for rectangles but handles arbitrary shapes`
        - check each cardinal `neighbor` of `point` in `direction`.  if `neighbor` is wall, check the neighbor in `direction` for a floor.  If that's a floor, and its not part of an area in the disjoint set union containing `area`, then `neighbor` is a potential connector
- Proceed as normal.  As you join areas, union the sets representing them in the disjoint set.

Basically the same algorithm, just treating the areas as arbitrarily-shaped.

I'd have to do some practical testing, to determine 1. if the above functions functionally equivalently and 2. that it does not function too slowly

**Comment by Chris3606** [comment]
After experimentation -- a complete algorithmic fix that results in perfect adherence to given parameters gets complicated.  Barring any additional ideas, for 1.8.2, the code change will be to just use `MapAreaFinder` to locate disconnected areas after connection , and connect them.  While the fix may interfere with 'maxSidesToConnect` being adhered to 100% of the time, it is a quick and viable fix for the mean-time.  If needed, a "nicer" fix (that could result in interface changes) can then be evaluated in 2.0.
